### PR TITLE
network: filter alt interfaces for binding by well-known prefixes

### PR DESCRIPTION
**Pull request by goaway**
Signed-off-by: Mike Schore <mike.schore@gmail.com>

### CODE REVIEW

**Comment by goaway** [comment]
cc @Reflejo

File: library/common/network/configurator.cc [hunk 1]
```diff
@@ -283,15 +294,31 @@ InterfacePair Configurator::getActiveAlternateInterface(envoy_network_t network,
     // WiFi should always support multicast, and will not be point-to-point.
     auto interfaces =
         enumerateInterfaces(family, IFF_UP | IFF_MULTICAST, IFF_LOOPBACK | IFF_POINTOPOINT);
-    return interfaces.size() > 0 ? interfaces[0] : std::make_pair("", nullptr);
+    for (const auto& interface : interfaces) {
+      // Look for interface with name that matches the expected prefix.
+      // TODO(goaway): This is quite brittle. It would be an improvement to:
+      //   1) Improve the scoping via flags.
+      //   2) Prioritize interfaces by prefix instead of simply filtering them.
+      if (std::get<const std::string>(interface).rfind(WlanPrefix, 0), 0) {
```
Comment (Reflejo) [inline library/common/network/configurator.cc:302]: was this supposed to be == 0?

Comment (goaway) [reply library/common/network/configurator.cc]: yep, thanks

File: library/common/network/configurator.cc [hunk 2]
```diff
@@ -283,15 +294,31 @@ InterfacePair Configurator::getActiveAlternateInterface(envoy_network_t network,
     // WiFi should always support multicast, and will not be point-to-point.
     auto interfaces =
         enumerateInterfaces(family, IFF_UP | IFF_MULTICAST, IFF_LOOPBACK | IFF_POINTOPOINT);
-    return interfaces.size() > 0 ? interfaces[0] : std::make_pair("", nullptr);
+    for (const auto& interface : interfaces) {
+      // Look for interface with name that matches the expected prefix.
+      // TODO(goaway): This is quite brittle. It would be an improvement to:
+      //   1) Improve the scoping via flags.
+      //   2) Prioritize interfaces by prefix instead of simply filtering them.
+      if (std::get<const std::string>(interface).rfind(WlanPrefix, 0) == 0) {
```
Comment (snowp) [inline library/common/network/configurator.cc:302]: I'd have used `absl::StartsWith`, but this works

File: library/common/network/configurator.cc [hunk 3]
```diff
@@ -53,6 +53,17 @@
...
 #endif
 
+#if defined(__APPLE__)
+constexpr absl::string_view WlanPrefix = "en";
+constexpr absl::string_view WwanPrefix = "pdp_ip";
+#elif defined(__ANDROID_API__)
+constexpr absl::string_view WlanPrefix = "wlan";
+constexpr absl::string_view WwanPrefix = "rmnet";
+#else
+constexpr absl::string_view WlanPrefix = "";
```
Comment (snowp) [inline library/common/network/configurator.cc:63]: I would include a comment here saying that this is effectively disable filtering, not immediately clear from reading the code.

**Review by snowp — APPROVED** [review]
LGTM with a few small comments. Tests would be nice but sounds like we need to be able to mock out the syscalls first in order to get that

File: library/common/network/configurator.cc (hunk 2 above)
Comment (jpsim) [reply library/common/network/configurator.cc:302]: `absl::StartsWith` would have been a lot easier to understand this code as an inexperienced C++ programmer.

Comment (goaway) [reply library/common/network/configurator.cc]: Updated to use absl::StartsWith for clarity.

**Review by snowp — APPROVED** [review]
LGTM, thanks!

**Pull request merged**